connector/src/main/java/com/google/cloud/hive/bigquery/connector/output/indirect/IndirectUtils.java

```diff
@@ -0,0 +1,111 @@
+ * limitations under the License.
+ */
+package com.google.cloud.hive.bigquery.connector.output.indirect;
+
+import com.google.cloud.bigquery.TableId;
+import com.google.cloud.hive.bigquery.connector.Constants;
+import com.google.cloud.hive.bigquery.connector.config.HiveBigQueryConfig;
+import com.google.cloud.hive.bigquery.connector.utils.HiveUtils;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageException;
+import com.google.cloud.storage.StorageOptions;
+import java.io.IOException;
+import java.util.List;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import org.apache.commons.lang.StringUtils;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapred.TaskAttemptID;
+import repackaged.by.hivebqconnector.com.google.common.collect.ImmutableList;
+
+public class IndirectUtils {
+
+  static Pattern gcsUriPattern = Pattern.compile("gs://([^/]*)(.*)?");
+
+  /** Retrieves the bucket name from a fully-qualified GCS URI. */
+  public static String extractBucketNameFromGcsUri(String gcsURI) {
+    Matcher m = gcsUriPattern.matcher(gcsURI);
+    if (m.find()) {
+      return m.group(1);
+    } else {
+      throw new RuntimeException("Incorrect GCS URI: " + gcsURI);
+    }
+  }
+
+  /** Returns true if the logged-in user has access to the given GCS path. */
+  public static boolean hasGcsWriteAccess(String gcsURI) {
+    String bucket = extractBucketNameFromGcsUri(gcsURI);
+    Storage storage = StorageOptions.newBuilder().build().getService();
+    List<Boolean> booleans;
+    try {
+      booleans = storage.testIamPermissions(bucket, ImmutableList.of("storage.objects.create"));
+    } catch (StorageException e) {
+      return false;
+    }
+    return !booleans.contains(false);
+  }
+
+  /**
+   * Returns the name of the temporary working directory in GCS where the temporary Avro files will
+   * be stored.
+   */
+  public static Path getGcsTempDir(Configuration conf, String gcsPathBase) {
+    String base = StringUtils.removeEnd(gcsPathBase, "/");
+    return new Path(
+        String.format(
+            "%s/%s%s",
+            base,
+            conf.get(
+                HiveBigQueryConfig.WORK_DIR_NAME_PREFIX_KEY,
+                HiveBigQueryConfig.WORK_DIR_NAME_PREFIX_DEFAULT),
+            HiveUtils.getHiveId(conf)));
+  }
+
+  /** Deletes the directory on GCS that contains the temporary Avro files for the job. */
+  public static void deleteGcsTempDir(Configuration conf, String gcsPathBase) throws IOException {
+    Path dir = getGcsTempDir(conf, gcsPathBase);
+    FileSystem fs = dir.getFileSystem(conf);
+    if (fs.exists(dir)) {
+      fs.deleteOnExit(dir);
+    }
+  }
+
+  /** Return the name prefix for the temp Avro file. */
+  public static String getTaskTempAvroFileNamePrefix(TableId tableId) {
+    return String.format(
```

Comment:
Can you please add a random string? like System.nanotime()

You would want of course to cache this value...

Reply:
This is the prefix, but the actual full name would include the task ID (see the function `getTaskAvroTempFile()`), which would make the file unique. Is that good enough?